File: scripts/api-audit/crossref.py
#!/usr/bin/env python3
"""Cruza llamadas frontend vs paths backend.

Normaliza templates del frontend (/sales/orders/${id}/ -> /sales/orders/{param}/)
y los empareja contra los paths del schema. Detecta llamadas sin match.

Uso: python3 crossref.py [backend.json] [frontend.json] [salida.json]
"""
import json
import re
import sys
from collections import defaultdict

backend_path_arg = sys.argv[1] if len(sys.argv) > 1 else "/tmp/opencode/backend-paths.json"
frontend_path_arg = sys.argv[2] if len(sys.argv) > 2 else "/tmp/opencode/frontend-calls.json"
out_path = sys.argv[3] if len(sys.argv) > 3 else "/tmp/opencode/crossref.json"

backend = json.load(open(backend_path_arg))
calls = json.load(open(frontend_path_arg))

def strip_interpolation(t):
    out = []
    i = 0
    while i < len(t):
        if t[i] == "$" and i + 1 < len(t) and t[i + 1] == "{":
            depth = 1
            i += 2
            start = i
            while i < len(t) and depth:
                if t[i] == "{":
                    depth += 1
                elif t[i] == "}":
                    depth -= 1
                i += 1
            inner = t[start : i - 1]
            if inner.strip() == "API_BASE":
                continue
            is_query = (
                re.search(r"[`'\"]\?", inner)
                or re.fullmatch(r"\s*(q|qs|query|queries|params|searchParams|filterQuery|search|firstQs)\s*", inner)
            )
            if not is_query:
                out.append("{param}")
        else:
            out.append(t[i])
            i += 1
    return "".join(out)

def normalize(t):
    t = strip_interpolation(t)
    t = re.sub(r"\{param\}\?.*$", "{param}", t)
    t = re.sub(r"\?.*$", "", t)
    if not t.startswith("/"):
        t = "/" + t
    if not t.startswith("/api/"):
        t = "/api" + t
    return t

backend_paths = set(backend.keys())
by_segments = defaultdict(list)
for p in backend_paths:
    n = len([s for s in p.strip("/").split("/") if s])
    by_segments[n].append(p)

def signature(path):
    return tuple("{p}" if re.fullmatch(r"\{[^}]+\}", s) else s for s in path.strip("/").split("/"))

sig_index = defaultdict(list)
for p in backend_paths:
    sig_index[signature(p)].append(p)

def match(path_t):
    for candidate in (path_t, re.sub(r"/\{param\}$", "", path_t)):
        sig = signature(candidate)
        for cand in sig_index.get(sig, []):
            return cand
        if candidate in backend_paths:
            return candidate
        rx = re.escape(candidate).replace(re.escape("{param}"), r"[^/]+")
        rx = rx.rstrip("/") + "/?"
        segs = len([s for s in candidate.strip("/").split("/") if s])
        for dn in (0, 1):
            for cand in by_segments.get(segs + dn, []):
                if re.fullmatch(rx, cand):
                    return cand
    return None

used = {}
unmatched = []
used_ops = {}
for c in calls:
    t = normalize(c["template"])
    bp = match(t)
    ref = f"{c['file']}:{c['line']}"
    if bp is None:
        unmatched.append({**c, "normalized": t})
    else:
        used.setdefault(bp, []).append(f"{c['method']} {ref}")
        key = f"{c['method']} {bp}"
        used_ops[key] = used_ops.get(key, 0) + 1

unused = {}
for p, methods in backend.items():
    for m in methods:
        key = f"{m} {p}"
        if key not in used_ops:
            unused.setdefault(p, []).append(m)

json.dump({
    "used_paths": used,
    "used_ops": used_ops,
    "unmatched_frontend_calls": unmatched,
    "unused_backend": unused,
    "stats": {
        "backend_paths": len(backend),
        "backend_ops": sum(len(v) for v in backend.values()),
        "frontend_calls": len(calls),
        "backend_paths_used": len(used),
        "backend_ops_used": len(used_ops),
        "backend_ops_unused": sum(len(v) for v in unused.values()),
        "unmatched_frontend": len(unmatched),
    },
}, open(out_path, "w"), indent=1, ensure_ascii=False)

s = json.load(open(out_path))
print(json.dumps(s["stats"], indent=2))
print("\n-- UNMATCHED FRONTEND CALLS --")
for u in unmatched:
    print(f"  {u['method']:6} {u['normalized'][:90]}  <- {u['file']}:{u['line']}")

File: scripts/api-audit/crossref.mjs
#!/usr/bin/env node
/**
 * Cruza las llamadas API del frontend contra el schema OpenAPI del backend Yggdra.
 * Por defecto lee el freeze ../yggdra_infra/api/schema/openapi.yaml.
 * Override: YGGDRA_SCHEMA_YML o primer argumento CLI.
 * Reporta llamadas frontend sin endpoint en el schema y drift de método HTTP.
 * Exit 1 si hay gaps (para usar como gate).
 * Uso: npm run audit:api  [ruta/al/openapi.yaml]
 */

import { readFileSync, readdirSync } from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { load as yamlLoad } from "js-yaml";

const REPO_ROOT = path.resolve(path.dirname(fileURLToPath(import.meta.url)), "../..");
const SRC_ROOT = path.join(REPO_ROOT, "src");
const SCHEMA = path.resolve(
  process.argv[2] ?? process.env.YGGDRA_SCHEMA_YML ?? path.join(REPO_ROOT, "../yggdra_infra/api/schema/openapi.yaml"),
);

const spec = yamlLoad(readFileSync(SCHEMA, "utf8"));
const backend = {};
for (const [p, methods] of Object.entries(spec.paths || {}).sort()) {
  const entry = {};
  for (const [m, op] of Object.entries(methods)) {
    if (!["get", "post", "put", "patch", "delete"].includes(m)) continue;
    entry[m.toUpperCase()] = { tags: op.tags || [] };
  }
  if (Object.keys(entry).length) backend[p] = entry;
}

const FNAMES = ["apiFetch", "apiFile", "apiGet", "apiPost", "apiPut", "apiPatch", "apiDelete", "fetch"];
const fnRe = new RegExp("\\b(" + FNAMES.join("|") + ")\\b", "g");
const methodRe = /\bmethod\s*:\s*["'](GET|POST|PUT|PATCH|DELETE)["']/;

function readString(src, i) {
  const quote = src[i];
  const out = [];
  i++;
  while (i < src.length) {
    const ch = src[i];
    if (ch === "\\") { out.push(src.slice(i, i + 2)); i += 2; continue; }
    if (quote === "`" && ch === "$" && src[i + 1] === "{") {
      let depth = 1;
      const start = i;
      i += 2;
      while (i < src.length && depth) { if (src[i] === "{") depth++; else if (src[i] === "}") depth--; i++; }
      out.push(src.slice(start, i));
      continue;
    }
    if (ch === quote) return [out.join(""), i + 1];
    out.push(ch);
    i++;
  }
  return [out.join(""), i];
}

function callExtent(src, openParen) {
  let depth = 0, inStr = null, i = openParen;
  while (i < src.length) {
    const ch = src[i];
    if (inStr) {
      if (ch === "\\") { i += 2; continue; }
      if (ch === inStr) inStr = null;
      i++;
      continue;
    }
    if (ch === '"' || ch === "'" || ch === "`") inStr = ch;
    else if (ch === "(") depth++;
    else if (ch === ")") { depth--; if (depth === 0) return src.slice(openParen, i + 1); }
    i++;
  }
  return src.slice(openParen, openParen + 2000);
}

const calls = [];
function walk(dir) {
  for (const e of readdirSync(dir, { withFileTypes: true })) {
    const p = path.join(dir, e.name);
    if (e.isDirectory()) walk(p);
    else if (/\.(ts|tsx)$/.test(e.name)) {
      const src = readFileSync(p, "utf8");
      let m;
      fnRe.lastIndex = 0;
      while ((m = fnRe.exec(src))) {
        const fname = m[1];
        let i = m.index + fname.length;
        while (i < src.length && /[A-Za-z0-9_<>,.\s]/.test(src[i])) i++;
        if (i >= src.length || src[i] !== "(") continue;
        let j = i + 1;
        while (j < src.length && /[ \t\n]/.test(src[j])) j++;
        if (j >= src.length || !"\"'`".includes(src[j])) continue;
        const [tmpl] = readString(src, j);
        const lineNo = src.slice(0, m.index).split("\n").length;
        const extent = callExtent(src, i);
        let method;
        if (fname === "apiFetch" || fname === "fetch") {
          const mm = extent.match(methodRe);
          method = mm ? mm[1] : "GET";
        } else if (fname === "apiFile") method = "FILE";
        else method = fname.replace("api", "").toUpperCase();
        calls.push({ file: path.relative(SRC_ROOT, p).split("\\").join("/"), line: lineNo, call: fname, method, template: tmpl.replace(/\n/g, " ").slice(0, 200) });
      }
    }
  }
}
walk(SRC_ROOT);

function stripInterpolation(t) {
  const out = [];
  let i = 0;
  while (i < t.length) {
    if (t[i] === "$" && t[i + 1] === "{") {
      let depth = 1;
      i += 2;
      const start = i;
      while (i < t.length && depth) { if (t[i] === "{") depth++; else if (t[i] === "}") depth--; i++; }
      const inner = t.slice(start, i - 1);
      if (inner.trim() === "API_BASE") continue;
      const isQuery = /^\s*(q|qs|query|queries|params|searchParams|filterQuery|search|firstQs)\s*$/.test(inner);
      if (!isQuery) out.push("{param}");
    } else { out.push(t[i]); i++; }
  }
  return out.join("");
}

function normalize(t) {
  t = stripInterpolation(t).replace(/\{param\}\?.*$/, "{param}").replace(/\?.*$/, "");
  if (!t.startsWith("/")) t = "/" + t;
  if (!t.startsWith("/api/")) t = "/api" + t;
  return t;
}

const backendPaths = Object.keys(backend);
const segs = (p) => p.split("/").filter(Boolean).length;
const sig = (p) => p.split("/").filter(Boolean).map((s) => (/^\{[^}]+\}$/.test(s) ? "{p}" : s)).join("/");
const sigIndex = {};
for (const p of backendPaths) { const s = sig(p); (sigIndex[s] = sigIndex[s] || []).push(p); }
const esc = (s) => [...s].map((ch) => (/[a-zA-Z0-9]/.test(ch) ? ch : "\\" + ch)).join("");

function match(pathT) {
  for (const candidate of [pathT, pathT.replace(/\/\{param\}$/, "")]) {
    const s = sig(candidate);
    if (sigIndex[s]) return sigIndex[s][0];
    if (backend[candidate]) return candidate;
    const rx = new RegExp("^" + esc(candidate).replaceAll(esc("{param}"), "[^/]+").replace(/\/+$/, "") + "/?$");
    for (const dn of [0, 1]) for (const cand of backendPaths) { if (segs(cand) === segs(candidate) + dn && rx.test(cand)) return cand; }
  }
  return null;
}

const used = {};
const unmatched = [];
for (const c of calls) {
  const t = normalize(c.template);
  const bp = match(t);
  const ref = c.file + ":" + c.line;
  if (bp === null) unmatched.push({ ...c, normalized: t });
  else (used[bp] = used[bp] || []).push(c.method + " " + ref);
}

const drift = [];
for (const c of calls) {
  const bp = match(normalize(c.template));
  if (bp && c.method !== "FILE" && !(c.method in backend[bp]))
    drift.push(c.method + " " + normalize(c.template) + " <- " + c.file + ":" + c.line + "  backend: " + Object.keys(backend[bp]).join(","));
}

console.log(JSON.stringify({
  schema: path.relative(REPO_ROOT, SCHEMA),
  backend_paths: backendPaths.length,
  backend_ops: Object.values(backend).reduce((a, v) => a + Object.keys(v).length, 0),
  frontend_calls: calls.length,
  frontend_files: new Set(calls.map((c) => c.file)).size,
  backend_paths_used: Object.keys(used).length,
  unmatched_frontend: unmatched.length,
  method_drift: drift.length,
}, null, 2));

if (unmatched.length) {
  console.log("\n-- UNMATCHED FRONTEND CALLS --");
  for (const u of unmatched) console.log("  " + u.method.padEnd(6) + " " + u.normalized.slice(0, 100).padEnd(102) + " <- " + u.file + ":" + u.line);
}
if (drift.length) {
  console.log("\n-- METHOD DRIFT --");
  drift.forEach((d) => console.log("  " + d));
}
if (unmatched.length || drift.length) {
  console.error(`\n[audit:api] FALLO: ${unmatched.length} llamadas sin match, ${drift.length} drift de método.`);
  process.exit(1);
}
console.log("\n[audit:api] OK: todas las llamadas frontend resuelven contra el schema.");
